### ISSUE TITLE
[Help Wanted] strange behavior with chains stuck in a non-optimal region of the tree topology space

### ISSUE THREAD

**Issue by igoryanovich** [revbayes/revbayes#431]
**Describe the desired behavior.**
I've tried to replicate a MrBayes analysis from doi:10.1163/22105832-bja10022 in RevBayes. The MB .nex file from the original paper is attached, as is my .rev recasting. 

**What is going wrong?**
MB very quickly finds the optimal-likelihood segment of the tree topology space. RB behaves strangely: after the short burnin, each chain remains in the same likelihood region, see Tracer_RB_runs.png below. 

RWTY comparison with MB chains shows that the **best** RB chains move in the same region as the MB chains, but the RB chains stuck at a lower likelihood. See rb_vs_mb.pdf attached. 

**Anything else you've tried?**
I made sure to use a lot of SPR moves per generation; switched from simple MCMC to MC^3, trying deltaHeat=0.1 and 0.2, which did not help to get RB unstuck; with deltaHeat=0.5, hardly any changes between the chains were achieved (although I have not investigated that setting for a long time). 

In MB, more tree moves are implemented than in RB (unless I'm missing something): in addition to NNI and SPR (they use "extended SPR" - I'm not sure if the regular SPR is meant, or something slightly different), MB also uses TBR and parsimony-biased SPR. I've tried to switch those moves off. The switching off of ParsSPR changes MB's behavior noticeably, see mb_all_variants.pdf below, compare the first four plots from chains w/o ParsSPR and the other four, with ParsSPR moves. However, this is not the only reason MB and RB are different, because even without ParsSPR, MrBayes finds the highest-likelihood segment which RB chains miss, cf. rb_vs_mb.pdf. 

**Screenshots**
<img width="1200" alt="Tracer_RB_runs" src="https://github.com/revbayes/revbayes/assets/31929063/e69087f5-ed2e-4b6c-b81e-f8397614c1e7">
[mb_all_variants.pdf](https://github.com/revbayes/revbayes/files/14451975/mb_all_variants.pdf)
[rb_vs_mb.pdf](https://github.com/revbayes/revbayes/files/14451976/rb_vs_mb.pdf)

**Computer info**
I tried this both on the Vienna VSC, and on my MacOS Laptop. 

VSC: 
RevBayes version (1.1.1)
Build from  () on Tue Mar 28 19:44:44 CEST 2023

laptop: 
RevBayes version (1.2.1)
Build from tags/v1.2.1 (04666d) on Mon Nov  7 15:45:19 UTC 2022

**Additional context**
Attaching analysis files - delete the added extension .txt

Original MB analysis file by Gunnink et al.: [BantuMBgammaNY.nex.txt](https://github.com/revbayes/revbayes/files/14452058/BantuMBgammaNY.nex.txt)

My modification switching off ParsSPR and TBR moves: [BantuMBgammaNY_wo_pSPR_TBR.nex.txt](https://github.com/revbayes/revbayes/files/14452057/BantuMBgammaNY_wo_pSPR_TBR.nex.txt)

My RevBayes recoding: [SBa_MC3_for_issue.rev.txt](https://github.com/revbayes/revbayes/files/14452059/SBa_MC3_for_issue.rev.txt)

The Nexus file with the data - identical to the data part of the analysis file from Gunnink et al. [SBantu-GunninkEtAl2023.nex.txt](https://github.com/revbayes/revbayes/files/14452060/SBantu-GunninkEtAl2023.nex.txt)

**Comment by igoryanovich** [comment]
Something looks definitely wrong here: attaching the likelihood trace for four rather long parallel chains 
<img width="1200" alt="four_longer_chains" src="https://github.com/revbayes/revbayes/assets/31929063/3c54c12a-7950-49d7-aca1-1e287be2e8cd">

(Before the chains start, there was a short burnin period, via mc3 burnin function.)

**Comment by davidcerny** [comment]
Perhaps @jsigao might know more?

**Comment by zikhoury** [comment]
Can confirm this is happening with other datasets as well (severity seems to be data-dependent) when using dnUniformTopology() with mvNNI and mvSPR(). When doing clock and non-clock analyses for a given dataset, MCMC performance is often comparatively much better when using dnBDP() with mvFNPR(), mvGPR(), and mvNarrow().

**Comment by wrightaprilm** [comment]
Hi folks,  

I played around on this last night. There are a couple issues I ran into:


- The branch length model was different between the RB and MB files. I've attached an MB file that has the model specified in the RevBayes file.
- The pi values of the F81 model were fixed in the RB file. I'm attaching my edited file with moves placed on pi.

[brlensp.nex.txt](https://github.com/user-attachments/files/16321532/brlensp.nex.txt)

[piMoves.Rev.txt](https://github.com/user-attachments/files/16321533/piMoves.Rev.txt)

With these two issues fixed, the two programs now appear to sample the same areas of solution space. 

One other issue I spotted is that the restriction site model is excluding 18 of your characters for not being compatible with variable site coding. If you want to replicate this behavior in RB (you don't have to!), the command:

```
, coding="variable"
```

To your call to PhyloCTMC. Leaving this out will not apply the variable site correction and will use all the available data.

However, the MCMC chain is much less efficient in RB. The MB analysis converges within about 4 hours and wasn't close in RB after I let it run overnight. But that's something I know much less about, so I'll pipe down ;)

**Comment by igoryanovich** [comment]
Hi April, 

Thank you VERY much! 
It’s a big relief to learn that the problems were largely due to my mistakes when recoding, and not something bad with RB. Phew! 

All the best,
 Igor

**Comment by igoryanovich** [comment]
Hi zikhoury, 

Thank you for the suggestions about the alternative strategies for picking moves!

**Comment by davidcerny** [comment]
@igoryanovich Just to clarify, what @zikhoury is describing wouldn't just be a different move-picking strategy but an entirely different analysis: to be able to use `mvFNPR` instead of `mvSPR`, for example, you'd have to estimate clock trees (`TimeTree` objects) rather than nonclock trees (`BranchLengthTree` objects), with all the attendant changes in analytical choices (e.g., placing a joint prior on topologies and divergence times rather than on topologies alone, specifying some sort of clock model, etc.).

**Comment by zikhoury** [comment]
@davidcerny @igoryanovich Absolutely, and apologies for any resulting confusion. I was hoping to provide an additional point of reference that might narrow down the issue (something happening with analyses using nonclock trees but not other tree objects) rather than give suggestions on how to improve a specific real-world analysis.

A while back, I was running different types of analyses on the same sets of simulated sequence data. Regardless of whether nonclock trees or timetrees were used for simulation, there were often MCMC issues in inference runs using nonclock trees, but no issues when using timetrees (specifically birth-death). The BD runs worked as expected, but nonclock runs sometimes had different replicate chains sampling different regions of tree space, often not visiting the true topology at all.

My experience also coincides with @wrightaprilm regarding efficiency versus MrBayes (using the same or similar model) on the same data. So I suspect there might be something going on (or at least some inefficiency) specifically with proposals on nonclock trees in RevBayes.

**Comment by brpetrucci** [comment]
This seems to be resolved (outside of RevBayes being less efficient than MrBayes, which is a whole other can of worms), so I'm closing the issue
